### ISSUE TITLE
Objective function mismatch detected incorrectly sometimes

### ISSUE THREAD

**Issue by dan-blanchard** [EducationalTestingService/skll#179]
While @aoifecahill was nice enough to add the objective function mismatch detection in #174, unfortunately there were a couple issues that slipped by my review.  The main problem is that there are really four types of objective functions (as we list [here](https://skll.readthedocs.org/en/latest/run_experiment.html#tuning):
1.  Classification only
2.  Regression or classification with integer classes
3.  Regression or classification with binary classes
4.  Regression only

Currently all the objectives are lumped into just two groups: regression and classification.  Therefore, if you try to use one of the class 2 or 3 objectives with a classifier (rather than a regressor), it raises an exception.  That's definitely a bug.

I know scikit-learn has some utility functions for determining the "task type" from the training data and labels, but that will include tasks that we don't currently support (like unsupervised learning), so I think maybe we should just build our own for this one.  It should be pretty straightforward to look at the labels and decide based on check their types:
-  If they're all floats, only class 4 objectives are valid with a regressor.  
-  If they're ints, class 1 and 2 are valid for classifiers, and 2-4 are valid for regressors.  
-  If there are two labels, class 3 objectives are valid for classifiers regardless of the type of the label.
-  If all of the labels are strings, only class 1 objectives are valid (with a classifier).
